x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -140,6 +194,108 @@ public <T extends TransportResponse> void sendRequest(
+        return new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Transport.Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                if (shouldSendWithRemoteAccessHeaders(connection, request)) {
+                    sendWithRemoteAccessHeaders(connection, action, request, options, handler);
+                } else {
+                    // Send regular request, without remote access headers
+                    try {
+                        sender.sendRequest(connection, action, request, options, handler);
+                    } catch (Exception e) {
+                        handler.handleException(new SendRequestTransportException(connection.getNode(), action, e));
+                    }
+                }
+            }
+
+            private boolean shouldSendWithRemoteAccessHeaders(final Transport.Connection connection, final TransportRequest request) {
+                final Authentication authentication = securityContext.getAuthentication();
+                assert authentication != null : "authentication must be present in security context";
+                final Subject effectiveSubject = authentication.getEffectiveSubject();
+                final Optional<String> remoteClusterAlias = remoteClusterAliasResolver.apply(connection);
+                return remoteClusterAlias.isPresent()
+                    && remoteClusterAuthorizationResolver.resolveAuthorization(remoteClusterAlias.get()) != null
+                    && isWhitelistedForRemoteAccessHeaders(request)
+                    && effectiveSubject.getType().equals(Subject.Type.USER)
+                    && false == User.isInternal(effectiveSubject.getUser())
+                    && Arrays.stream(effectiveSubject.getUser().roles()).noneMatch(ReservedRolesStore::isReserved);
+            }
+
+            private <T extends TransportResponse> void sendWithRemoteAccessHeaders(
+                final Transport.Connection connection,
+                final String action,
+                final TransportRequest request,
+                final TransportRequestOptions options,
+                final TransportResponseHandler<T> handler
+            ) {
+                logger.debug("Sending request with remote access headers for [{}] action", action);
+
+                if (connection.getVersion().before(VERSION_REMOTE_ACCESS_HEADERS)) {
+                    handler.handleException(
+                        new TransportException(
+                            "Settings for remote cluster indicate remote access headers should be sent but target cluster version ["
+                                + connection.getVersion()
+                                + "] does not support receiving them"
```

Comment:
Yup, exactly my thinking -- throwing surfaces the configuration error explicitly. 

To pass it to the handler, it needs to be a `TransportException` (or subclass thereof). I can wrap it in a `SendRequestTransportException` here, so something like `new SendRequestTransportException(..., new IllegalArgumentException(...))`.

Reply:
Can we just throw the exception? My understanding is that TransportService does the wrapping anyway (or if we consider this is composed inside another interceptor, it will be wrapped by `sendWithUser`). Directly calling `handler.handleException` is useful when we need proactively restore threadContext before handling exception. But in this case, we haven't modified threadContext yet.

Reply:
Ah yes for sure, we can throw. My thinking around passing the exception to the handler is still from back when this logic was in a standalone interceptor. But since this is not an internal implementation detail of the overall sec server interceptor it makes even more sense to throw here and let the exception get wrapped upstream.